---
title: "QTL_Heatmap"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DOQTL)
#use new QTLHeatmap function
qtl.heatmap = function (lod, chr=NULL, borders = TRUE, ...) 
{
  old.par = par(no.readonly = TRUE)
  # Strip off the location data.
  hdr = lod[, 1:3]
  lod = as.matrix(lod[, -1:-3])
  # Get the chr boundaries and midpoints.
  chrbnd = table(hdr$Chr)
  chrbnd = chrbnd[order(as.numeric(names(chrbnd)))]
  chrbnd = cumsum(chrbnd)
  chrmid = c(0, chrbnd[-length(chrbnd)]) + diff(c(0, chrbnd)) * 0.5
  # Exponentiate the data and then scale each phenotype between 0 and 1.
  lod = exp(lod)
  lod = lod/matrix(apply(lod, 2, max, na.rm = TRUE), nrow(lod),
                   ncol(lod), byrow = TRUE)
  # Cluster the LOD profiles.
  lod.cor = cor(lod, use = "pair")
  lod.cl = hclust(as.dist(1 - lod.cor), method = "average")
  lod = lod[, lod.cl$order]
  dend = as.dendrogram(lod.cl)
  # Create the heatmap.
  layout(matrix(1:2, 1, 2), widths = c(0.1, 0.9))
  par(plt = c(0, 1, 0.1, 0.9))
  plot(dend, horiz = TRUE, axes = FALSE, yaxs = "i", leaflab = "none")
  par(las = 1, plt = c(0, 0.8, 0.1, 0.9))
  breaks = 0:100/100
  col = colorRampPalette(c(gray(10:2/10), rgb(1, 0.5, 0), rgb(1, 0, 0)))(length(breaks) - 1)
  if(!is.null(chr)){
    lod=subset(lod, hdr$Chr==chr)
    hdr=subset(hdr, hdr$Chr==chr)}
  image(1:nrow(lod), 1:ncol(lod), lod, breaks = breaks, col = col, 
        ann = FALSE, axes = FALSE)
  #add the borders if TRUE
  if(borders){
    if(!is.null(chr)){
      abline(h = 0:ncol(lod) + 0.5, col = "grey30")
    }else{
      #only add vertical borders if plotting whole genome
      abline(v = c(0, chrbnd), col = "grey30")
      abline(h = 0:ncol(lod) + 0.5, col = "grey30")
    }
  }
  mtext(text = colnames(lod), side = 4, at = 1:ncol(lod))
  #add chr # one x-axis if plotting whole genome
  if(is.null(chr)){
    mtext(names(chrbnd), side = 1, line = 0, at = chrmid, font = 2)
   # mtext(names(chrbnd), side = 3, line = 0, at = chrmid, font = 2)
  }else{
    #add cM locations if plotting single chr
    cM=seq(0, ceiling(max(hdr$cM)), by=20)
    mtext(cM, side = 1, line = 0, at = seq(nrow(hdr)*0.02,nrow(hdr), by=nrow(hdr)/length(cM)), font = 2)
    # mtext(cM, side = 3, line = 0, at = seq(nrow(hdr)*0.02,nrow(hdr), by=nrow(hdr)/length(cM)), font = 2)
  }
  # Legend along bottom.
  width = nrow(lod)/length(col)
  pin = par("pin")
  usr = par("usr")
  mai = par("mai")
  usr.per.in = (usr[4] - usr[3])/pin[2]
  top = -mai[1] * 0.25 * usr.per.in
  bottom = top - usr.per.in * 0.25
  par(xpd = NA)
  for (i in 1:length(col)) {
    rect((i - 1) * width, bottom, i * width, top, col = col[i], 
         border = NA)
  }
  par(old.par)
}
```

## Use the QTL Heatmap to Find Common QTLs in PC1 Phenotypes 
load data
```{r load}
setwd("~/Projects/DO")
#load in the phenotype scanone data
load("results/7-26-16/Svenson_pheno_add.RData", verbose=T)
load("results/7-26-16/Svenson_pheno_intdiet.RData", verbose=T)
load("results/7-26-16/Svenson_pheno_intsex.RData", verbose=T)
load("data/Svenson_4eQTLMapping.Rdata", verbose=T)

#read the phenotypes within Principle Component 1 and 2 
pc1=read.csv("results/7-20-16/pc1.csv")
pc2=read.csv("results/7-20-16/pc2.csv")
```

## Plot the full genome heatmap
```{r whole}
##PC1 Phenotypes with borders
ind=colnames(pheno.scan1.add$lod)%in%pc1$X
lod=cbind(snps[,c(1:2,4)], as.data.frame(pheno.scan1.add$lod[,ind]))
#scale bar below indicate the relative LOD score, look for darkgrey/red vertical lines
qtl.heatmap(lod)
title("PC1 phenotypes | Additive Covariates")
##
##PC2 Phenotypes without borders
ind=colnames(pheno.scan1.add$lod)%in%pc2$X
lod2=cbind(snps[,c(1:2,4)], as.data.frame(pheno.scan1.add$lod[,ind]))
qtl.heatmap(lod2, borders = FALSE)
title("PC2 phenotypes | Additive Covariates")

```

##Plot individual chromosomes
(Don't trust the cM just yet, they need to be fixed)
```{r single chr}
#PC1 chr6
qtl.heatmap(lod, chr=6)
title("PC1 phenotypes on Chr6 | Additive Covariates")

#PC1 chr9 without borders
qtl.heatmap(lod, chr=9, border=FALSE)
title("PC1 phenotypes on Chr9 | Additive Covariates")

#PC2 chr1 without borders
qtl.heatmap(lod2, chr=1, borders=FALSE)
title("PC1 phenotypes on Chr6 | Additive Covariates")
```